src/language/patterns.md

```diff
@@ -0,0 +1,362 @@
+}                    
+```
+
+To ignore parts of a matched value, you can use a [wildcard pattern][]
+as a place holder. In the case of list patterns, you can use a [rest element][].
+
+### Destructuring
+
+When an object and pattern match, the pattern can then access the object's data 
+and extract it in parts. In other words, the pattern _destructures_ the object:
+
+```dart
+var numList = [1, 2, 3];
+// List pattern [a, b, c] destructures the three elements from numList...
+var [a, b, c] = numList;
+// ...and assigns them to new variables.
+print(a + b + c);          
+```
+
+You can nest [any kind of pattern][types] inside a destructuring pattern. 
+For example, this case pattern matches and destructures a two-element
+list whose first element is `a` or `b`:
+
+```dart
+switch (list) {
+  case ['a' || 'b', var c]: 
+  // ...
+}
+```
+
+## Places patterns can appear
+
+You can use patterns in several places in the Dart language:
+
+<a id="pattern-uses"></a>
+
+- Local variable [declarations](#variable-declaration) and [assignments](#variable-assignment)
+- [for and for-in loops][for]
+- [if-case][if] and [switch-case][switch]
+- Control flow in [collection literals][]
+
+This section describes common use cases for matching and destructuring with patterns.
+
+### Variable declaration
+
+You can use a _pattern variable declaration_ anywhere Dart allows local variable
+declaration. 
+The pattern matches against the value on the right of the declaration.
+Once matched, it destructures the value and binds it to new local variables:
+
+```dart
+// Declares new variables a, b, and c.
+var (a, [b, c]) = ("str", [1, 2]);
+```
+
+A pattern variable declaration must start with either `var` or `final`, followed
+by a pattern. 
+
+### Variable assignment 
+
+A _variable assignment pattern_ falls on the left side of an assignment.
+First, it destructures the matched object. Then it assigns the values to
+_existing_ variables, instead of binding new ones. 
+
+Use a variable assignment pattern to swap the values of two variables without
+declaring a third temporary one:
+
+```dart
+var (a, b) = ('left', 'right');
+(b, a) = (a, b); // Swap.
+print('$a $b'); // Prints "right left".
+```
+
+### Switch statements and expressions
+
+Every case clause contains a pattern. This applies to [switch statements][switch]
+and [expressions][], as well as [if-case statements][if].
+You can use [any kind of pattern][types] in a case.
+
+_Case patterns_ are [refutable][].
+They allow control flow to either:
+- Match and destructure the object being switched on.
+- Continue execution if the object doesn't match.
+
+```dart
+switch (obj) {
+  case 1: ...
+    // Matches if 1 == obj.
+  case >= first && <= last: ...  
+    // Matches if the value of obj is between the constant values of 'first' and 'last'.
+  case (var a, var b): ...
+    // Matches if obj is a record with two fields, then assigns the fields to 'a' and 'b'.
+  case Rect(width: var w, height: var h): ...
```

Comment:
Shouldn't those `var` properties be `final` as it is a Dart's best practice to declare final whenever the is no point change a property? Am I missing something? 👀

Reply:
There's a balance in documenting Dart, as both ways are supported and `var` is shorter, which is especially beneficial in this case. 

Unlike on Flutter, different Dart terms prefer different ways, as long as you are consistent.

_I do prefer `final` personally, but sometimes `var` is more approachable for the docs._